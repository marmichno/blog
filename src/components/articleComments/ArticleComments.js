import React from 'react';
import { useEffect, useState } from 'react';
import ArticleCommentsCSS from './articleComments.module.css';
import {useSelector} from 'react-redux';
import {getArticleComments} from '../../requests/getArticleComments';
import {postArticleComment} from "../../requests/postArticleComment";

export const ArticleComments = () => {

    const [comments, setComments] = useState([]);
    const selectedArticleId = useSelector(state => state.chooseArticleReducer);

    const [userName, setUserName] = useState("");
    const [userEmail, setUserEmail] = useState("");
    const [userComment, setUserComment] = useState("");

    useEffect(() => {
        getComments();
    },[]);

    const getComments = async () => {
        const response = await getArticleComments(selectedArticleId);
        setComments(response);
    }

    const postComment = async () => {
        const request = await postArticleComment(selectedArticleId, userName, userEmail, userComment);
        getComments();
    }

    //inputs
    const handleUserCommentChange = (e) => {
        setUserComment(e.target.value);
    }

    const handleUserEmailChange = (e) => {
        setUserEmail(e.target.value);
    }

    const handleUserNameChange = (e) => {
        setUserName(e.target.value);
    }   

    if(comments !== undefined){
        return(
            <div className={ArticleCommentsCSS.commentsContainer}>
                <div className={ArticleCommentsCSS.commentsContainer__postCommentContainer}>
                    <label htmlFor="userName" className={ArticleCommentsCSS.commentsContainer__postCommentContainer__label}>
                        <span className={ArticleCommentsCSS.commentsContainer__postCommentContainer__span}>
                            <p className={ArticleCommentsCSS.commentsContainer__postCommentContainer__span__paragraph}>Name</p>
                        </span>
                    </label>
                    
                    <input type="text" autoComplete="off" name="userName" 
                    className={ArticleCommentsCSS.commentsContainer__postCommentContainer__input}
                    onChange = {handleUserNameChange}
                    required>
                    </input>

                    <label htmlFor="userEmail" className={ArticleCommentsCSS.commentsContainer__postCommentContainer__label}>
                        <span className={ArticleCommentsCSS.commentsContainer__postCommentContainer__span}>
                            <p className={ArticleCommentsCSS.commentsContainer__postCommentContainer__span__paragraph}>Email</p>
                        </span>
                    </label>

                    <input type="text" autoComplete="off" name="userEmail" 
                    className={ArticleCommentsCSS.commentsContainer__postCommentContainer__input}
                    onChange = {handleUserEmailChange}
                    required>
                    </input>

                    <label htmlFor="commentBody" className={ArticleCommentsCSS.commentsContainer__postCommentContainer__label}>
                        <span className={ArticleCommentsCSS.commentsContainer__postCommentContainer__span}>
                            <p className={ArticleCommentsCSS.commentsContainer__postCommentContainer__span__paragraph}>Comment</p>
                        </span>
                    </label>

                    <textarea type="text" autoComplete="off" name="commentBody" 
                    className={ArticleCommentsCSS.commentsContainer__postCommentContainer__textArea}
                    onChange = {handleUserCommentChange}
                    required>
                    </textarea>
                    <button className={ArticleCommentsCSS.commentsContainer__postCommentContainer__button} onClick={postComment}>Add comment</button>
                </div>

                {comments.map(value => {
                    return <div className={ArticleCommentsCSS.commentsContainer__commentContainer}>
                        <p className={ArticleCommentsCSS.commentsContainer__commentContainer__name}>{value.name} 
                            <span className={ArticleCommentsCSS.commentsContainer__commentContainer__name__date}> - 07/12/2021 14:26</span>
                        </p>
                        <p className={ArticleCommentsCSS.commentsContainer__commentContainer__email}>{value.email}</p>
                        <p className={ArticleCommentsCSS.commentsContainer__commentContainer__body}>{value.body}</p>
                    </div>
                })}
            </div>
        )
    }else{
        return(
            <div className={ArticleCommentsCSS.commentsContainer}></div>
        )
    }
}