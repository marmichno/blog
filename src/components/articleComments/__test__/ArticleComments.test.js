import { createStore, applyMiddleware} from 'redux';
import { Provider } from 'react-redux';
import {ArticleComments} from '../ArticleComments';
import validArticleComments from '../__mocks__/validArticleComments.json';
import {render, waitFor, fireEvent} from '@testing-library/react';
import React from 'react';
import "@testing-library/jest-dom/extend-expect";
import TestRenderer from "react-test-renderer";

const store = createStore(() => [], {}, applyMiddleware());

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve(validArticleComments),
  })
);

afterEach(() => {
  jest.clearAllMocks();
});

test("check if comments are visible after api call", async () => {

    const {getAllByText} = render(<Provider store={store}><ArticleComments/></Provider>);

    await waitFor(() => {
        expect(getAllByText('id labore ex et quam laborum'));
        expect(getAllByText('Eliseo@gardner.biz'));
        expect(getAllByText('laudantium enim quasi est quidem magnam voluptate ipsam eos tempora quo necessitatibus dolor quam autem quasi reiciendis et nam sapiente accusantium'));
    })
});

test("handles exception - renders without comments", async () => {

fetch.mockImplementationOnce(() => Promise.reject("API FAILURE"));

  const {getByTestId} = render(<Provider store={store}><ArticleComments/></Provider>);

  await waitFor(() => {
    expect(getByTestId("commentsContainer")).toBeTruthy();
  })
});

test("get request is called after post request", async () => {
  
    const {getByTestId} = render(<Provider store={store}><ArticleComments/></Provider>);

    await waitFor(() => {
      fireEvent.click(getByTestId("postRequest"));
  })
  
    await waitFor(() => {
      expect(fetch).toHaveBeenCalledTimes(3);
    })
  });

test("matches snapshot", () => {
  const tree = TestRenderer.create(<Provider store={store}><ArticleComments/></Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});