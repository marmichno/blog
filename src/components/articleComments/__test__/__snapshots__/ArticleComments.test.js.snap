// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="commentsContainer"
  data-testid="commentsContainer"
>
  <div
    className="commentsContainer__postCommentContainer"
  >
    <label
      className="commentsContainer__postCommentContainer__label"
      htmlFor="userName"
    >
      <span
        className="commentsContainer__postCommentContainer__span"
      >
        <p
          className="commentsContainer__postCommentContainer__span__paragraph"
        >
          Name
        </p>
      </span>
    </label>
    <input
      autoComplete="off"
      className="commentsContainer__postCommentContainer__input"
      name="userName"
      onChange={[Function]}
      required={true}
      type="text"
    />
    <label
      className="commentsContainer__postCommentContainer__label"
      htmlFor="userEmail"
    >
      <span
        className="commentsContainer__postCommentContainer__span"
      >
        <p
          className="commentsContainer__postCommentContainer__span__paragraph"
        >
          Email
        </p>
      </span>
    </label>
    <input
      autoComplete="off"
      className="commentsContainer__postCommentContainer__input"
      name="userEmail"
      onChange={[Function]}
      required={true}
      type="text"
    />
    <label
      className="commentsContainer__postCommentContainer__label"
      htmlFor="commentBody"
    >
      <span
        className="commentsContainer__postCommentContainer__span"
      >
        <p
          className="commentsContainer__postCommentContainer__span__paragraph"
        >
          Comment
        </p>
      </span>
    </label>
    <textarea
      autoComplete="off"
      className="commentsContainer__postCommentContainer__textArea"
      name="commentBody"
      onChange={[Function]}
      required={true}
      type="text"
    />
    <button
      className="commentsContainer__postCommentContainer__button"
      data-testid="postRequest"
      onClick={[Function]}
    >
      Add comment
    </button>
  </div>
</div>
`;
